src/framerate_gui.cpp

```diff
@@ -0,0 +1,813 @@
+			/* Always 2:1 graph area */
+			graph_size.width = 2 * graph_size.height;
+			*size = graph_size;
+
+			size->width += size_ms_label.width + 2;
+			size->height += size_s_label.height + 2;
+		}
+	}
+
+	void SelectHorizontalScale(TimingMeasurement range)
+	{
+		/* Determine horizontal scale based on period covered by 60 points
+		* (slightly less than 2 seconds at full game speed) */
+		struct ScaleDef { TimingMeasurement range; int scale; };
+		static const ScaleDef hscales[] = {
+			{ 120, 60 },
+			{  10, 20 },
+			{   5, 10 },
+			{   3,  4 },
+			{   1,  2 },
+		};
+		for (const ScaleDef *sc = hscales; sc < hscales + lengthof(hscales); sc++) {
+			if (range < sc->range) this->horizontal_scale = sc->scale;
+		}
+	}
+
+	void SelectVerticalScale(TimingMeasurement range)
+	{
+		/* Determine vertical scale based on peak value (within the horizontal scale + a bit) */
+		static const TimingMeasurement vscales[] = {
+			TIMESTAMP_PRECISION * 100,
+			TIMESTAMP_PRECISION * 10,
+			TIMESTAMP_PRECISION * 5,
+			TIMESTAMP_PRECISION,
+			TIMESTAMP_PRECISION / 2,
+			TIMESTAMP_PRECISION / 5,
+			TIMESTAMP_PRECISION / 10,
+			TIMESTAMP_PRECISION / 50,
+			TIMESTAMP_PRECISION / 200,
+		};
+		for (const TimingMeasurement *sc = vscales; sc < vscales + lengthof(vscales); sc++) {
+			if (range < *sc) this->vertical_scale = (int)*sc;
+		}
+	}
+
+	/** Recalculate the graph scaling factors based on current recorded data */
+	void UpdateScale()
+	{
+		const TimingMeasurement *durations = _pf_data[this->element].durations;
+		const TimingMeasurement *timestamps = _pf_data[this->element].timestamps;
+		int num_valid = _pf_data[this->element].num_valid;
+		int point = _pf_data[this->element].prev_index;
+
+		TimingMeasurement lastts = timestamps[point];
+		TimingMeasurement time_sum = 0;
+		TimingMeasurement peak_value = 0;
+		int count = 0;
+
+		/* Sensible default for when too few measurements are available */
+		this->horizontal_scale = 4;
+
+		for (int i = 1; i < num_valid; i++) {
+			point--;
+			if (point < 0) point = NUM_FRAMERATE_POINTS - 1;
+
+			TimingMeasurement value = durations[point];
+			if (value == PerformanceData::INVALID_DURATION) {
+				/* Skip gaps in data by pretending time is continuous across them */
+				lastts = timestamps[point];
+				continue;
+			}
+			if (value > peak_value) peak_value = value;
+			count++;
+
+			/* Accumulate period of time covered by data */
+			time_sum += lastts - timestamps[point];
+			lastts = timestamps[point];
+
+			/* Enough data to select a range and get decent data density */
+			if (count == 60) this->SelectHorizontalScale(time_sum / TIMESTAMP_PRECISION);
+
+			/* End when enough points have been collected and the horizontal scale has been exceeded */
+			if (count >= 60 && time_sum >= (this->horizontal_scale + 2) * TIMESTAMP_PRECISION / 2) break;
+		}
+
+		this->SelectVerticalScale(peak_value);
+	}
+
+	virtual void OnTick()
+	{
+		this->SetDirty();
+
+		if (this->scale_update_timer > 0) {
```

Comment:
Considering basing this off real time instead of ticks, I think that might give better response times when performance is bad, and worse when performance is good.